api/v1beta1/openstackdataplane_webhook.go

```diff
@@ -0,0 +1,125 @@
+/*
+Copyright 2023.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	"fmt"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+const errNoSSHKey = "SSH key not provided for Node %s"
+
+// log is for logging in this package.
+var openstackdataplanelog = logf.Log.WithName("openstackdataplane-resource")
+
+// SetupWebhookWithManager sets up the webhook with the Manager
+func (r *OpenStackDataPlane) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// TODO(user): EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+
+//+kubebuilder:webhook:path=/mutate-dataplane-openstack-org-v1beta1-openstackdataplane,mutating=true,failurePolicy=fail,sideEffects=None,groups=dataplane.openstack.org,resources=openstackdataplanes,verbs=create;update,versions=v1beta1,name=mopenstackdataplane.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Defaulter = &OpenStackDataPlane{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *OpenStackDataPlane) Default() {
+	openstackdataplanelog.Info("default", "name", r.Name)
+	r.Spec.Default()
+
+}
+
+// Default - set defaults for this Glance spec
+func (spec *OpenStackDataPlaneSpec) Default() {
+	for nodeName, node := range spec.Nodes {
+		if node.HostName == "" {
+			node.HostName = nodeName
+		}
+		if node.Node.AnsibleSSHPrivateKeySecret == "" {
```

Comment:
So this copies the value to node level when the node is initially added, and updating the key reference later requires updating the field in each node.

I'm ok with this but we might want to revisit this later if we consider supporting key rotation for all nodes(though that can be done by easy sed replacement, I guess)

Reply:
yes I understand your point though it might be a bit confusing that the result can be different according to the API.

This would not block the change. We can discuss this further if we want. So I'll close this conversation now.

Reply:
if it's about using  PATCH to update the ssh keys as an use-case, we can probably implement PATCH for the webhook at some point.